stage_binaries.sh

```diff
@@ -0,0 +1,278 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# USAGE: stage-binaries.sh -o <staging-dir> ( -p <package> | -b binary )..."
+#
+# Stages all the packages or files and their dependencies (+ libraries and
+# copyrights) to the staging dir.
+#
+# This is intended to be used in a multi-stage docker build with a distroless/base
+# or distroless/cc image.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# A handler for when we exit automatically on an error.
+# Borrowed from kubernetes, which was borrowed from
+# https://gist.github.com/ahendrix/7030300
+function errexit() {
+  # If the shell we are in doesn't have errexit set (common in subshells) then
+  # don't dump stacks.
+  set +o | grep -qe "-o errexit" || return
+
+  local file="$(basename "${BASH_SOURCE[1]}")"
+  local line="${BASH_LINENO[0]}"
+  local func="${FUNCNAME[1]:-}"
+  echo "FATAL: error at ${func}() ${file}:${line}" >&2
+}
+
+# trap ERR to provide an error handler whenever a command exits nonzero  this
+# is a more verbose version of set -o errexit
+trap 'errexit' ERR
+
+# setting errtrace allows our ERR trap handler to be propagated to functions,
+# expansions and subshells
+set -o errtrace
+
+# file_to_package identifies the debian package that provided the file $1
+function file_to_package() {
+    local file="$1"
+
+    # `dpkg-query --search $file-pattern` outputs lines with the format: "$package: $file-path"
+    # where $file-path belongs to $package
+    # https://manpages.debian.org/jessie/dpkg/dpkg-query.1.en.html
+    dpkg-query --search "$(realpath "${file}")" | cut -d':' -f1
+}
+
+# package_to_copyright gives the path to the copyright file for the package $1
+function package_to_copyright() {
+    local pkg="$1"
+    echo "/usr/share/doc/${pkg}/copyright"
+}
+
+# stage_file stages the filepath $1 to $2, following symlinks
+# and staging copyrights
+function stage_file() {
+    local file="$1"
+    local staging="$2"
+
+    # copy the named path
+    cp -a --parents "${file}" "${staging}"
+
+    # recursively follow symlinks
+    if [[ -L "${file}" ]]; then
+        stage_file "$(cd "$(dirname "${file}")" || exit; realpath -s "$(readlink "${file}")")" "${staging}"
+    fi
+
+    # get the package so we can stage package metadata as well
+    local package="$(file_to_package "${file}")"
+    # stage the copyright for the file, if it exists
+    local copyright="$(package_to_copyright "${package}")"
+    if [[ -f "${copyright}" ]]; then
+        cp -a --parents "${copyright}" "${staging}"
+    fi
+
+    # stage the package status mimicking bazel
+    # https://github.com/bazelbuild/rules_docker/commit/f5432b813e0a11491cf2bf83ff1a923706b36420
+    # instead of parsing the control file, we can just get the actual package status with dpkg
+    dpkg -s "${package}" > "${staging}/var/lib/dpkg/status.d/${package}"
+}
+
+# helper because bash sometimes barfs on embedded comments
```

Comment:
nit: I can't really tell from this comment and the usage context why this is needed. At the call site it's also not obvious what's happening without going back and reading this.

Reply:
grep treats "no match" as an error, which trips up `pipefail` and `errexit`

```
$ echo foo | grep bar; echo $?
1
```

So we want to do:

```
$ echo foo | { grep bar || true; }; echo $?
0
```

Adding back the comments in the style of `binary_to_libraries()`:

```
$ echo foo \
    `# comment1` \
    | { grep "bar" || true; } \
    `# comment2`; echo $?
bash: syntax error near unexpected token ``# comment2`'
```
Parens fails the same way.

It works without the braces and `|| true`, but returns error status

```
$ echo foo     `# comment1`     | grep "bar"     `# comment2`; echo $?
1
```

Once bash parses `{}` or `()` it can't handle subsequent `#comment` blocks.  `happy()` hides that.  I could rename it :)

Reply:
cute, I guess I should actually read the man page :)

Reply:
PTAL.  I want to run a scanner against it to be sure it works, when I get a moment